### ISSUE TITLE
Composer-rest-server - Deployment with Docker : Error trying to ping. Error: REQUEST_TIMEOUT

### ISSUE THREAD

**Issue by maxgfr** [hyperledger/composer#4126]
Hello everyone,

## Context
I tried to follow this tutorial : https://hyperledger.github.io/composer/latest/integrating/deploying-the-rest-server

## Problem
At the end of the tutorial, I had an error when I launch `docker logs -f rest`  :

```shell
[2018-06-09 12:57:16] PM2 log: Launching in no daemon mode
[2018-06-09 12:57:16] PM2 log: Starting execution sequence in -fork mode- for app name:composer-rest-server id:0
[2018-06-09 12:57:16] PM2 log: App name:composer-rest-server id:0 online
Discovering types from business network definition ...
Connection fails: Error: Error trying to ping. Error: REQUEST_TIMEOUT
It will be retried for the next request.
Exception: Error: Error trying to ping. Error: REQUEST_TIMEOUT
Error: Error trying to ping. Error: REQUEST_TIMEOUT
    at _checkRuntimeVersions.then.catch (/home/composer/.npm-global/lib/node_modules/composer-rest-server/node_modules/composer-connector-hlfv1/lib/hlfconnection.js:790:34)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:228:7)
[2018-06-09 13:02:21] PM2 log: App [composer-rest-server] with id [0] and pid [14], exited with code [1] via signal [SIGINT]
[2018-06-09 13:02:21] PM2 log: Starting execution sequence in -fork mode- for app name:composer-rest-server id:0
[2018-06-09 13:02:21] PM2 log: App name:composer-rest-server id:0 online
Discovering types from business network definition ...
```

## Expected Behavior
The command `docker logs -f rest` should finish with `Browse your REST API at http://localhost:3000/explorer`

## Actual Behavior
Get the error as detailed above

## Possible Fix
I tried changing the permissions of ~/.composer by using the command below but it didn't change the output: `chmod -R 777 ~/.composer`

## My Environment
* composer-cli and composer-rest-server 0.19.6
* node 8.11.2

**Comment by rthatcher** [comment]
Please add the exact command used to start the REST container.

**Note** that this error can be seen when the connection.json file for the card starting the rest server refers to _localhost_ in the URLs for the fabric server containers.

**Comment by maxgfr** [comment]
Hi,

Thanks for your answer ! 

- Here is my Dockerfile :
```shell
FROM hyperledger/composer-rest-server
RUN npm install --production loopback-connector-mongodb passport-github && \
    npm cache clean --force && \
    ln -s node_modules .node_modules
```

- To start my container : 
```shell
sudo docker run \
    -d \
    -e COMPOSER_CARD=${COMPOSER_CARD} \
    -e COMPOSER_NAMESPACES=${COMPOSER_NAMESPACES} \
    -e COMPOSER_AUTHENTICATION=${COMPOSER_AUTHENTICATION} \
    -e COMPOSER_MULTIUSER=${COMPOSER_MULTIUSER} \
    -e COMPOSER_DATASOURCES="${COMPOSER_DATASOURCES}"\
    -e NODE_APP_INSTANCE=""\
    -v ~/.composer:/home/composer/.composer \
    --name rest \
    --network composer_default \
    -p 3000:3000 \
    myorg/my-composer-rest-server
```

- Before starting my container : 
```shell
export COMPOSER_CARD=admin@magma-network
export COMPOSER_NAMESPACES=always
export COMPOSER_AUTHENTICATION=false
export COMPOSER_MULTIUSER=false
export COMPOSER_DATASOURCES=$(<datasources.json)
```

- My file datasources.json 
```shell
{
    "db": {
        "name": "db",
        "connector": "mongodb",
        "host": "mongo"
    }
}
```

- My connection.json  : 
```shell
{
    "name": "magma-network",
    "x-type": "hlfv1",
    "version": "1.0.0",
    "peers": {
        "peer0.org1.example.com": {
            "url": "grpc://localhost:7051",
            "eventUrl": "grpc://localhost:7053"
        }
    },
    "certificateAuthorities": {
        "ca.org1.example.com": {
            "url": "http://localhost:7054",
            "caName": "ca.org1.example.com"
        }
    },
    "orderers": {
        "orderer.example.com": {
            "url": "grpc://localhost:7050"
        }
    },
    "organizations": {
        "Org1": {
            "mspid": "Org1MSP",
            "peers": [
                "peer0.org1.example.com"
            ],
            "certificateAuthorities": [
                "ca.org1.example.com"
            ]
        }
    },
    "channels": {
        "composerchannel": {
            "orderers": [
                "orderer.example.com"
            ],
            "peers": {
                "peer0.org1.example.com": {
                    "endorsingPeer": true,
                    "chaincodeQuery": true,
                    "eventSource": true
                }
            }
        }
    },
    "client": {
        "organization": "Org1",
        "connection": {
            "timeout": {
                "peer": {
                    "endorser": "300",
                    "eventHub": "300",
                    "eventReg": "300"
                },
                "orderer": "300"
            }
        }
    }
}
```

**Comment by hingjlee** [comment]
I encountered the same issue even though I changed respective 'localhost' addresses with docker hostnames and set up restadmin business card. 

`{
  "error": {
    "statusCode": 500,
    "name": "Error",
    "message": "Error trying to ping. Error: Failed to connect before the deadline",
    "stack": "Error: Error trying to ping. Error: Failed to connect before the deadline\n    at _checkRuntimeVersions.then.catch (/home/composer/.npm-global/lib/node_modules/composer-rest-server/node_modules/composer-connector-hlfv1/lib/hlfconnection.js:806:34)\n    at <anonymous>"
  }
}`

I can successfully authenticate through Google and can import wallets but cannot connect to any transaction. Importing wallet is interacting with mongo db and it seems loopback setting is correct. The reason of the error is the 'rest' docker cannot communicate with the 'frabic' docker.

`{"name":"hlfv1","x-type":"hlfv1","x-commitTimeout":300,"version":"1.0.0","client":{"organization":"Org1","connection":{"timeout":{"peer":{"endorser":"300","eventHub":"300","eventReg":"300"},"orderer":"300"}}},"channels":{"composerchannel":{"orderers":["orderer.example.com"],"peers":{"peer0.org1.example.com":{}}}},"organizations":{"Org1":{"mspid":"Org1MSP","peers":["peer0.org1.example.com"],"certificateAuthorities":["ca.org1.example.com"]}},"orderers":{"orderer.example.com":{"url":"grpc://orderer.example.com:7050"}},"peers":{"peer0.org1.example.com":{"url":"grpc://peer0.org1.example.com:7051"}},"certificateAuthorities":{"ca.org1.example.com":{"url":"http://ca.org1.example.com:7054","caName":"ca.org1.example.com"}}}`

The 'sed' command successfully replace 'localhost' but still cannot talk to the 'fabrics' in docker. Please help. Thanks.

**Comment by rthatcher** [comment]
I think you may have a different problem `Error: Failed to connect before the deadline\n at _checkRuntimeVersions.then.catch`

Have a look at this Q&A on Stack Overflow https://stackoverflow.com/questions/51875604/deploying-a-hyperledger-composer-blockchain-business-network-to-hyperledger-fabr/51875874

**Comment by hingjlee** [comment]
Thank you very much for the quick responses. The connection is fine as both in CLI and playground, i can connect to the network and do transactions.

**Comment by hingjlee** [comment]
Hi Rob, I would like to provide extra info. I can successfully using [Enabling authentication for the REST server](https://hyperledger.github.io/composer/latest/integrating/enabling-rest-authentication) to enable everything. In this section, no 'loopback' is required but however I prefer using Google+ rather than github as it's more common. 

`nohup composer-rest-server -c admin@biz-network -m true -a true -w true & disown`

I used above common to launch the service but it does consider the high availability section as mentioned in the original article. 

I also tried to install google-oauth2 not in docker and tried to execute to avoid the docker issue. In this trial, I use the ordinary business card instead of the restadmin special connection.json card.

`npm install -g passport-google-oauth2`

However, in this trial, I got another error.

**Comment by rthatcher** [comment]
This is a new problem, not related to the original issue. 
Can I suggest that you raise this as a Stack Overflow question rather than here (it is more likely to be seen and answered too)
BTW have you seen this tutorial https://hyperledger.github.io/composer/latest/tutorials/google_oauth2_rest

**Comment by hingjlee** [comment]
Hi Rob, I'll raise a new issue. My whole issue is following the Google_OAuth2 tutorial. As stated in first comment, the issue I found is that the docker rest-server/loopback cannot talk to the fabric docker instance even though I changed "localhost" to respective docker names.

**Comment by hingjlee** [comment]
Hi Rob, I teared down the whole setup and redo it again. I tried to use your suggestion.

`COMPOSER_MULTIUSER=false`

I further created queries.qry which does not have any permissions attached. I can successfully query the blockchain under MULTIUSER mode=false. 

However, when I switched back to MULTIUSER mode, "Error trying to ping" error persists.
